### PR TITLE
NEW #5711 Add shipment line deleting and editing for draft shipments.

### PR DESCRIPTION
**Pull request by fappels**
# New #5711 Add shipment line deleting and editing for draft shipments.
- Update class for line actions
- Add lot form select
- Add delete button to delete a line
- Add edit button to:
      - edit qty
      - select a new source location
      - select other lot number
      - edit line extrafield

### CODE REVIEW

File: htdocs/expedition/class/expeditionbatch.class.php [hunk 1]
```diff
@@ -207,6 +210,7 @@ static function fetchAll($db,$id_line_expdet)
 				$tmp->fk_origin_stock = $obj->fk_origin_stock;
 				$tmp->fk_expeditiondet = $obj->fk_expeditiondet;
 				$tmp->dluo_qty = $obj->qty;
+				$tmp->entrepot_id = $obj->fk_entrepot;
```
Comment (eldy) [inline htdocs/expedition/class/expeditionbatch.class.php]: This data come from a link that is ok just after the reception is done, but as soon as stock is empty, the line into product_batch and product_stock will no exists anymore and we won't be able to get the entrepot_id value for old reception. Is this ok for you ?

File: htdocs/product/class/html.formproduct.class.php [hunk 2]
```diff
@@ -329,5 +330,126 @@ function load_measuring_units($name='measuring_units', $measuring_style='', $def
...
+		dol_syslog(get_class($this).'::loadLotStock', LOG_DEBUG);
+		$resql = $this->db->query($sql);
+		if ($resql)
+		{
+			$num = $this->db->num_rows($resql);
+			$i = 0;
+			while ($i < $num)
+			{
+				$obj = $this->db->fetch_object($resql);
+				$this->cache_lot[$obj->rowid]['id'] =$obj->rowid;
```
Comment (eldy) [inline htdocs/product/class/html.formproduct.class.php]: If fk_product or fk_entrepot are set at first call, we load data restricted to product or warehouse. And the cache array contains only this restricted data.
If we wall again the method loadLotStock with empty fk_product and empty fk_entrepot, then it return the cache. But this cache does not contains all data, so content is not complete for the second call. I suggest to add a key into cache
$this->cache_lot[fk_product.'_'.fk_warehouse]

File: htdocs/expedition/class/expeditionbatch.class.php (hunk 1 above)
Comment (fappels) [reply htdocs/expedition/class/expeditionbatch.class.php]: I will revert this, thought I need the entrepot_id to easy get the warehouse source of the lot ready to ship, but I seem to have it resolved it in another way. I get it now with a fetch from product_batch. I need the warehouse from product_batch to get it's source location, you should not remove stock before validating. If you remove the lot from stock you also won't be able to validate the shipment. This is also an advantage of the new edit option. When you get the error the lot isn't available anymore, you can replace it by another lot for validating.

File: htdocs/product/class/html.formproduct.class.php (hunk 2 above)
Comment (fappels) [reply htdocs/product/class/html.formproduct.class.php]: I will check this, but will need to consider memory/performance. A company can have a lot off different lot/serial numbers in stock.

**Comment by BebZ** [comment]
"When you get the error the lot isn't available anymore, you can replace it by another lot for validating."
this is such a major improvement, Francis is the new warehouse's guru :)
"Add shipment line deleting and editing for draft shipments"
> this is exactly the way a shipment must be edited, in a draft it's perfect. 
Just 2 cents in a comment : 
After validation, it means the products left the warehouse (so must be invoiced),
so it should be corrected when still in draft, shipment deletion should be really exceptionnal and recorded in blockchain (regarding fiscal regulations and "validations" of versions regarding laws to respected, in France).

File: htdocs/product/class/html.formproduct.class.php (hunk 2 above)
Comment (fappels) [reply htdocs/product/class/html.formproduct.class.php]: I have added a new parameter to be able to cash all lot stock for all products in lines. So you could also use this new form select for creating lines. I don't see any use case of loading all lot numbers for all products.

**Pull request merged**